---

title: Title

keywords: fastai
sidebar: home_sidebar



---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 02_tasks.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%</span><span class="k">load_ext</span> autoreload
<span class="o">%</span><span class="k">autoreload</span> 2
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="task_get_file" class="doc_header"><code>task_get_file</code><a href="https://github.com/btrspg/create_reference/tree/master/create_reference/tasks.py#L19" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>task_get_file</code>(<strong><code>link</code></strong>, <strong><code>filename</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="task_bwa_build_index" class="doc_header"><code>task_bwa_build_index</code><a href="https://github.com/btrspg/create_reference/tree/master/create_reference/tasks.py#L29" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>task_bwa_build_index</code>(<strong><code>software</code></strong>, <strong><code>reference</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="task_bowtie_build_index" class="doc_header"><code>task_bowtie_build_index</code><a href="https://github.com/btrspg/create_reference/tree/master/create_reference/tasks.py#L37" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>task_bowtie_build_index</code>(<strong><code>software</code></strong>, <strong><code>reference</code></strong>, <strong><code>prefix</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="task_bowtie2_build_index" class="doc_header"><code>task_bowtie2_build_index</code><a href="https://github.com/btrspg/create_reference/tree/master/create_reference/tasks.py#L46" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>task_bowtie2_build_index</code>(<strong><code>software</code></strong>, <strong><code>reference</code></strong>, <strong><code>prefix</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="task_samtools_build_index" class="doc_header"><code>task_samtools_build_index</code><a href="https://github.com/btrspg/create_reference/tree/master/create_reference/tasks.py#L55" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>task_samtools_build_index</code>(<strong><code>software</code></strong>, <strong><code>reference</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="task_hisat2_build_index" class="doc_header"><code>task_hisat2_build_index</code><a href="https://github.com/btrspg/create_reference/tree/master/create_reference/tasks.py#L62" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>task_hisat2_build_index</code>(<strong><code>software</code></strong>, <strong><code>reference</code></strong>, <strong><code>prefix</code></strong>, <strong><code>snp</code></strong>=<em><code>''</code></em>, <strong><code>gtf</code></strong>=<em><code>''</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="task_gunzip" class="doc_header"><code>task_gunzip</code><a href="https://github.com/btrspg/create_reference/tree/master/create_reference/tasks.py#L85" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>task_gunzip</code>(<strong><code>gzipfile</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">bwa</span><span class="o">=</span><span class="s1">&#39;bwa&#39;</span>
<span class="n">bowtie</span><span class="o">=</span><span class="s1">&#39;bowtie&#39;</span>
<span class="n">bowtie2</span><span class="o">=</span><span class="s1">&#39;bowtie2&#39;</span>
<span class="n">hisat2</span><span class="o">=</span><span class="s1">&#39;hisat2&#39;</span>
<span class="n">reference</span> <span class="o">=</span> <span class="s1">&#39;genome.fasta&#39;</span>
<span class="n">bowtie_prefix</span> <span class="o">=</span> <span class="s1">&#39;/tmp/bowtie_index&#39;</span>
<span class="n">bowtie2_prefix</span><span class="o">=</span><span class="s1">&#39;/tmp/bowtie2_index&#39;</span>
<span class="nb">print</span><span class="p">(</span><span class="n">task_bwa_build_index</span><span class="p">(</span><span class="n">bwa</span><span class="p">,</span><span class="n">reference</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">task_bowtie_build_index</span><span class="p">(</span><span class="n">bowtie</span><span class="p">,</span><span class="n">reference</span><span class="p">,</span><span class="n">bowtie_prefix</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">task_bowtie2_build_index</span><span class="p">(</span><span class="n">bowtie2</span><span class="p">,</span><span class="n">reference</span><span class="p">,</span><span class="n">bowtie2_prefix</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>bwa  index  genome.fasta
mkdir -p /tmp
bowtie-build    genome.fasta /tmp/bowtie_index
mkdir -p /tmp
bowtie2-build    genome.fasta /tmp/bowtie2_index
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">task_hisat2_build_index</span><span class="p">(</span><span class="n">hisat2</span><span class="p">,</span><span class="n">reference</span><span class="p">,</span><span class="s1">&#39;/tmp/hisat2_prefix&#39;</span><span class="p">,</span><span class="s1">&#39;xxxx.snp&#39;</span><span class="p">,</span><span class="s1">&#39;xxx.gtf&#39;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">task_hisat2_build_index</span><span class="p">(</span><span class="n">hisat2</span><span class="p">,</span><span class="n">reference</span><span class="p">,</span><span class="s1">&#39;/tmp/hisat2_prefix&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">,</span><span class="s1">&#39;xxx.gtf&#39;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">task_hisat2_build_index</span><span class="p">(</span><span class="n">hisat2</span><span class="p">,</span><span class="n">reference</span><span class="p">,</span><span class="s1">&#39;/tmp/hisat2_prefix&#39;</span><span class="p">,</span><span class="s1">&#39;xxxx.snp&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>mkdir -p /tmp
awk &#39;BEGIN{OFS=&#34;\t&#34;} {if($2 ~ /^chr/) {$2 = substr($2, 4)}; if($2 == &#34;M&#34;) {$2 = &#34;MT&#34;} print}&#39; xxxx.snp \
    &gt; /tmp/hisat2_prefix_snp.tmp
hisat2_extract_snps_haplotypes_UCSC.py genome.fasta xxxx.snp /tmp/hisat2_prefix
hisat2_extract_splice_sites.py xxx.gtf &gt; /tmp/hisat2_prefix.ss
hisat2_extract_exons.py xxx.gtf &gt; /tmp/hisat2_prefix.exon
hisat2-build -p  16  genome.fasta  --ss /tmp/hisat2_prefix.ss --exon /tmp/hisat2_prefix.exon --haplotype /tmp/hisat2_prefix.haplotype --genome_snp /tmp/hisat2_prefix.snp /tmp/hisat2_prefix
mkdir -p /tmp
hisat2_extract_splice_sites.py xxx.gtf &gt; /tmp/hisat2_prefix.ss
hisat2_extract_exons.py xxx.gtf &gt; /tmp/hisat2_prefix.exon
hisat2-build -p  16  genome.fasta  --ss /tmp/hisat2_prefix.ss --exon /tmp/hisat2_prefix.exon /tmp/hisat2_prefix
mkdir -p /tmp
awk &#39;BEGIN{OFS=&#34;\t&#34;} {if($2 ~ /^chr/) {$2 = substr($2, 4)}; if($2 == &#34;M&#34;) {$2 = &#34;MT&#34;} print}&#39; xxxx.snp \
    &gt; /tmp/hisat2_prefix_snp.tmp
hisat2_extract_snps_haplotypes_UCSC.py genome.fasta xxxx.snp /tmp/hisat2_prefix
hisat2-build -p  16  genome.fasta  --haplotype /tmp/hisat2_prefix.haplotype --genome_snp /tmp/hisat2_prefix.snp /tmp/hisat2_prefix
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">hs_link</span><span class="o">=</span><span class="s1">&#39;ftp://ftp.ensembl.org/pub/release-99/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz&#39;</span>
<span class="n">ss_link</span><span class="o">=</span><span class="s1">&#39;ftp://ftp.ensembl.org/pub/release-99/fasta/sus_scrofa/dna/Sus_scrofa.Sscrofa11.1.dna.toplevel.fa.gz&#39;</span>
<span class="n">outdir</span><span class="o">=</span><span class="s1">&#39;/tmp/&#39;</span>
<span class="n">species</span><span class="o">=</span><span class="s1">&#39;homo_sapiens&#39;</span>
<span class="n">version</span><span class="o">=</span><span class="s1">&#39;99&#39;</span>
<span class="n">filename</span><span class="o">=</span><span class="s1">&#39;genome.fasta.gz&#39;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">task_get_file</span><span class="p">(</span><span class="n">hs_link</span><span class="p">,</span><span class="n">filename</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>echo No need to make dirs
wget -c ftp://ftp.ensembl.org/pub/release-99/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz -O genome.fasta.gz
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">task_get_file</span><span class="p">(</span><span class="n">ss_link</span><span class="p">,</span><span class="n">filename</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>echo No need to make dirs
wget -c ftp://ftp.ensembl.org/pub/release-99/fasta/sus_scrofa/dna/Sus_scrofa.Sscrofa11.1.dna.toplevel.fa.gz -O genome.fasta.gz
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

